File: engine/dispatch.py
"""Pure orchestration, split into plan() and execute() so a GitHub-specific
adapter can commit state to the cron-state branch *between* them (claim
committed before anything executes, finalize committed after) — see
adapters/github/run_dispatcher.py. Neither function touches git, files
outside a StateBackend, or anything else GitHub-specific.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Optional

from . import scheduler
from .concurrency import run_bounded
from .executors.base import run_with_retry
from .history import HistoryStore
from .ledger import Ledger
from .models import Job, RunResult
from .notify import Notifier
from .timeutil import iso


class DispatchError(Exception):
    pass


@dataclass
class PlannedRun:
    now: datetime
    to_run: list[tuple[Job, datetime, bool]] = field(default_factory=list)
    evaluated_job_ids: set = field(default_factory=set)
    prior_failures: dict = field(default_factory=dict)
    skipped_count: int = 0
    stale_claims_reconciled: list = field(default_factory=list)


@dataclass
class DispatchSummary:
    now: datetime
    ran: list[RunResult] = field(default_factory=list)
    skipped_count: int = 0
    stale_claims_reconciled: list = field(default_factory=list)


def plan(
    jobs: list[Job],
    ledger: Ledger,
    history: HistoryStore,
    now: datetime,
    job_id: Optional[str] = None,
    force_disabled: bool = False,
) -> PlannedRun:
    """Figure out what's due and claim it. Nothing executes yet. Callers should
    persist state/dedup (and ideally commit it) right after calling this and
    before calling execute() — that's the safety boundary an interrupted run
    relies on."""
    stale = ledger.reconcile_stale_claims(now)

    to_run: list[tuple[Job, datetime, bool]] = []
    evaluated_job_ids: set = set()
    skipped_count = 0

    if job_id is not None:
        job = next((j for j in jobs if j.id == job_id), None)
        if job is None:
            raise DispatchError(f"unknown job id '{job_id}'")
        if not job.enabled and not force_disabled:
            raise DispatchError(
                f"job '{job_id}' is disabled — pass force_disabled=true to run it anyway"
            )
        to_run.append((job, now, True))
    else:
        for job in jobs:
            if not job.enabled:
                continue
            since = ledger.get_cursor(job.id, now)
            occurrences = scheduler.compute_due_occurrences(job.schedule, job.timezone, since, now)
            if not occurrences:
                continue
            evaluated_job_ids.add(job.id)
            run_list, skip_list = scheduler.apply_misfire_policy(
                occurrences, job.misfire_policy, job.misfire_cap
            )
            for occ in skip_list:
                history.append_skipped(
                    job.id, iso(occ), "skipped by misfire policy (a more recent occurrence ran instead)"
                )
                skipped_count += 1
            for occ in run_list:
                if ledger.is_claimed(job.id, occ):
                    continue
                to_run.append((job, occ, False))

    for job, occ, _manual in to_run:
        ledger.claim(job.id, occ, now)

    prior_failures = {job.id: ledger.consecutive_failures(job.id) for job, _occ, _m in to_run}

    return PlannedRun(
        now=now,
        to_run=to_run,
        evaluated_job_ids=evaluated_job_ids,
        prior_failures=prior_failures,
        skipped_count=skipped_count,
        stale_claims_reconciled=stale,
    )


def execute(
    planned: PlannedRun,
    ledger: Ledger,
    history: HistoryStore,
    notifier: Notifier,
    repo_root: Path,
    max_concurrency: int = 5,
    run_url: Optional[str] = None,
) -> DispatchSummary:
    """Run everything plan() claimed, record results, notify, advance cursors.

    `run_url` (if given) is attributed to every result from this tick — it's
    display metadata for the dashboard, not something that affects dedup,
    scheduling, or retries.
    """
    results = run_bounded(
        [(lambda j=job, o=occ: run_with_retry(j, o, repo_root)) for job, occ, _m in planned.to_run],
        max_workers=max_concurrency,
    )

    for (job, occ, manual), result in zip(planned.to_run, results):
        result.trigger = "manual" if manual else "scheduled"
        result.run_url = run_url
        ledger.finalize(job.id, occ, result.status, planned.now)
        history.append(job.id, result, job.history_limit)

        if not manual:
            ledger.advance_cursor(job.id, planned.now, result.status)

        was_failing = planned.prior_failures.get(job.id, 0) > 0
        if result.status == "failed" and job.notify_on_failure:
            notifier.notify_failure(job, result, ledger.consecutive_failures(job.id))
        elif result.status == "success" and was_failing and job.notify_on_recovery:
            notifier.notify_recovery(job, result)

    # Jobs evaluated this tick but left with nothing to run (e.g. everything
    # was already claimed) still need their watermark to move forward.
    ran_job_ids = {job.id for job, _occ, manual in planned.to_run if not manual}
    for jid in planned.evaluated_job_ids - ran_job_ids:
        last_status = ledger.cursors.get(jid, {}).get("last_status", "success")
        ledger.advance_cursor(jid, planned.now, last_status)

    ledger.prune_dedup(planned.now)
    ledger.record_heartbeat(
        planned.now, "ok", detail=f"{len(planned.to_run)} executed, {planned.skipped_count} skipped"
    )

    return DispatchSummary(
        now=planned.now,
        ran=results,
        skipped_count=planned.skipped_count,
        stale_claims_reconciled=planned.stale_claims_reconciled,
    )

File: src/engine/dispatch.js
"use strict";

const scheduler = require("./scheduler");
const { runBounded } = require("./concurrency");
const { runWithRetry } = require("./executors/base");
const { iso } = require("./timeutil");

class DispatchError extends Error {}

function plan(jobs, ledger, history, now, options = {}) {
  const { jobId = null, forceDisabled = false } = options;
  const stale = ledger.reconcileStaleClaims(now);
  const toRun = [];
  const evaluatedJobIds = new Set();
  let skippedCount = 0;

  if (jobId) {
    const job = jobs.find((candidate) => candidate.id === jobId);
    if (!job) throw new DispatchError(`unknown job id '${jobId}'`);
    if ((!job.enabled || ledger.isAutoDisabled(job.id)) && !forceDisabled) {
      throw new DispatchError(`job '${jobId}' is disabled - pass force_disabled=true to run it anyway`);
    }
    toRun.push({ job, occurrence: now, manual: true });
  } else {
    for (const job of jobs) {
      if (!job.enabled) continue;
      if (ledger.isAutoDisabled(job.id)) continue;
      const pauseUntil = ledger.failurePauseUntil(job.id);
      if (pauseUntil && pauseUntil > now) continue;
      const since = ledger.getCursor(job.id, now);
      const occurrences = scheduler.computeDueOccurrences(job.schedule, job.timezone, since, now);
      if (!occurrences.length) continue;
      evaluatedJobIds.add(job.id);
      const [runList, skipList] = scheduler.applyMisfirePolicy(
        occurrences,
        job.misfire_policy,
        job.misfire_cap
      );
      for (const occurrence of skipList) {
        history.appendSkipped(
          job.id,
          iso(occurrence),
          "skipped by misfire policy (a more recent occurrence ran instead)"
        );
        skippedCount += 1;
      }
      for (const occurrence of runList) {
        if (!ledger.isClaimed(job.id, occurrence)) {
          toRun.push({ job, occurrence, manual: false });
        }
      }
    }
  }

  for (const item of toRun) ledger.claim(item.job.id, item.occurrence, now);
  const priorFailures = {};
  for (const item of toRun) priorFailures[item.job.id] = ledger.consecutiveFailures(item.job.id);

  return {
    now,
    to_run: toRun,
    evaluated_job_ids: evaluatedJobIds,
    prior_failures: priorFailures,
    skipped_count: skippedCount,
    stale_claims_reconciled: stale,
  };
}

function failurePolicyPatch(job, planned, result) {
  if (result.status === "success") return {};
  const policy = job.failure_policy || {};
  const prior = planned.prior_failures[job.id] || 0;
  const failures = prior + 1;
  const initial = policy.initial_backoff_seconds ?? 300;
  const multiplier = policy.backoff_multiplier ?? 2;
  const max = policy.max_backoff_seconds ?? 21600;
  const autoDisableAfter = policy.auto_disable_after_consecutive_failures ?? 5;
  const backoffSeconds = Math.min(max, Math.round(initial * multiplier ** Math.max(0, failures - 1)));
  const patch = {
    consecutive_failures: failures,
    failure_pause_until_utc:
      backoffSeconds > 0 ? new Date(planned.now.getTime() + backoffSeconds * 1000).toISOString() : null,
    auto_disabled: false,
    auto_disabled_at_utc: null,
    auto_disabled_reason: null,
  };
  if (failures >= autoDisableAfter) {
    patch.auto_disabled = true;
    patch.auto_disabled_at_utc = planned.now.toISOString();
    patch.auto_disabled_reason = `${failures} consecutive failures`;
  }
  return patch;
}

async function execute(planned, ledger, history, notifier, repoRoot, options = {}) {
  const { maxConcurrency = 5, runUrl = null } = options;
  const tasks = planned.to_run.map(
    ({ job, occurrence }) => async () => runWithRetry(job, occurrence, repoRoot)
  );
  const results = await runBounded(tasks, maxConcurrency);

  for (let i = 0; i < planned.to_run.length; i += 1) {
    const { job, occurrence, manual } = planned.to_run[i];
    const result = results[i];
    result.trigger = manual ? "manual" : "scheduled";
    result.run_url = runUrl;
    ledger.finalize(job.id, occurrence, result.status, planned.now);
    history.append(job.id, result, job.history_limit);

    const safetyPatch = {
      ...failurePolicyPatch(job, planned, result),
      last_run_utc: result.finished_at,
      last_trigger: manual ? "manual" : "scheduled",
    };
    if (!manual) {
      ledger.advanceCursor(job.id, planned.now, result.status, safetyPatch);
    } else {
      ledger.recordJobHealth(job.id, planned.now, result.status, safetyPatch);
    }

    const wasFailing = (planned.prior_failures[job.id] || 0) > 0;
    if (result.status === "failed" && job.notify_on_failure) {
      await notifier.notifyFailure(job, result, ledger.consecutiveFailures(job.id));
    } else if (result.status === "success" && wasFailing && job.notify_on_recovery) {
      await notifier.notifyRecovery(job, result);
    }
  }

  const ranJobIds = new Set(planned.to_run.filter((item) => !item.manual).map((item) => item.job.id));
  for (const jobId of planned.evaluated_job_ids) {
    if (!ranJobIds.has(jobId)) {
      const lastStatus = (ledger.cursors[jobId] || {}).last_status || "success";
      ledger.advanceCursor(jobId, planned.now, lastStatus);
    }
  }

  ledger.pruneDedup(planned.now);
  const retentionDays = Math.max(...planned.to_run.map((item) => item.job.history_retention_days), 365);
  history.cleanupArchives(retentionDays, planned.now);
  ledger.recordHeartbeat(planned.now, "ok", `${planned.to_run.length} executed, ${planned.skipped_count} skipped`);

  return {
    now: planned.now,
    ran: results,
    skipped_count: planned.skipped_count,
    stale_claims_reconciled: planned.stale_claims_reconciled,
  };
}

module.exports = { DispatchError, plan, execute, failurePolicyPatch };
